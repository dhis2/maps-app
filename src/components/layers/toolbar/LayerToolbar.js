import React from 'react';
import PropTypes from 'prop-types';
import { Toolbar, ToolbarGroup } from 'material-ui/Toolbar';
import IconButton from 'material-ui/IconButton';
import IconMenu from 'material-ui/IconMenu';
import { SvgIcon } from '@dhis2/d2-ui-core';
import OpacitySlider from './OpacitySlider';
import DownloadMenu from './DownloadMenu';
import OpenAsMenu from './OpenAsMenu';

const styles = {
    toolbar: {
        backgroundColor: '#eee',
        height: 32,
        padding: '0 8px',
    },
    button: {
        float: 'left',
        padding: 4,
        width: 32,
        height: 32,
    },
    moreButton: {
        float: 'right',
        padding: 4,
        width: 32,
        height: 32,
        marginRight: -7,
        marginLeft: -5,
    },
    menuList: {
        paddingTop: 0,
        paddingBottom: 0,
    },
};

const OverlayToolbar = ({
    layer,
    onEdit,
    onRemove,
    toggleDataTable,
    onOpacityChange,
}) => (
    <Toolbar style={styles.toolbar}>
        <ToolbarGroup>
            {onEdit &&
                layer.type !== 'external' && (
                    <IconButton
                        onClick={() => onEdit(layer)}
                        tooltip="Edit"
                        tooltipPosition="top-center"
                        style={styles.button}
                    >
                        <SvgIcon icon="Create" />
                    </IconButton>
                )}
            {(layer.layer === 'thematic' ||
                layer.layer === 'boundary' ||
                layer.layer === 'facility') && (
                <IconButton
                    onClick={() => toggleDataTable(layer.id)}
                    tooltip="Data table"
                    tooltipPosition="top-center"
                    style={styles.button}
                >
                    <SvgIcon icon="ViewList" />
                </IconButton>
            )}

            <OpacitySlider
                {...layer}
                onChange={opacity => onOpacityChange(layer.id, opacity)}
            />
        </ToolbarGroup>

        <ToolbarGroup>
            {onRemove && (
                <IconButton
                    onClick={onRemove}
                    tooltip="Delete"
                    tooltipPosition="top-center"
                    style={styles.button}
                >
                    <SvgIcon icon="Delete" />
                </IconButton>
            )}

            {true === false && ( //    TODO
                <IconMenu
                    iconButtonElement={
                        <IconButton
                            tooltip="More"
                            tooltipPosition="top-center"
                            style={styles.moreButton}
                        >
                            <SvgIcon icon="MoreVert" />
                        </IconButton>
                    }
                    listStyle={styles.menuList}
                >
                    <OpenAsMenu {...layer} />
                    <DownloadMenu {...layer} />
                </IconMenu>
            )}
        </ToolbarGroup>
    </Toolbar>
);

OverlayToolbar.propTypes = {
    layer: PropTypes.object,
    onEdit: PropTypes.func,
    onRemove: PropTypes.func,
    onDataTableShow: PropTypes.func,
    onOpacityChange: PropTypes.func,
};

export default OverlayToolbar;
