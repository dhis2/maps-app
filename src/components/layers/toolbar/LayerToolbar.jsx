import i18n from '@dhis2/d2-i18n'
import { Tooltip, IconEdit24, IconView24, IconViewOff24 } from '@dhis2/ui'
import cx from 'classnames'
import PropTypes from 'prop-types'
import React from 'react'
import { IconButton } from '../../core/index.js'
import LayerToolbarMoreMenu from './LayerToolbarMoreMenu.jsx'
import OpacitySlider from './OpacitySlider.jsx'
import styles from './styles/LayerToolbar.module.css'

const LayerToolbar = ({
    hasOpacity = true,
    opacity = 1,
    isVisible,
    onOpacityChange,
    toggleLayerVisibility,
    hasError,
    ...expansionMenuProps
}) => {
    const onEdit = expansionMenuProps.onEdit

    return (
        <div className={styles.toolbar} data-test="layertoolbar">
            {onEdit && (
                <IconButton
                    tooltip={i18n.t('Edit')}
                    onClick={onEdit}
                    className={styles.editButton}
                    dataTest="layer-edit-button"
                >
                    <IconEdit24 />
                </IconButton>
            )}
            <IconButton
                tooltip={!hasError ? i18n.t('Toggle visibility') : null}
                onClick={toggleLayerVisibility}
                className={cx({
                    visible: isVisible, // for cypress testing only
                    notvisible: !isVisible, // for cypress testing only
                })}
                dataTest="visibilitybutton"
                disabled={hasError}
            >
                {isVisible ? <IconView24 /> : <IconViewOff24 />}
            </IconButton>

            <div className={styles.sliderContainer}>
                <Tooltip
                    content={
                        hasError
                            ? i18n.t('Layer is invalid')
                            : i18n.t('Set layer opacity')
                    }
                    disabled={hasError}
                >
                    <OpacitySlider
                        opacity={opacity}
                        disabled={hasError || !isVisible || !hasOpacity}
                        onChange={onOpacityChange}
                    />
                </Tooltip>
            </div>
            <div className={styles.menuButton}>
                <LayerToolbarMoreMenu
                    hasError={hasError}
                    {...expansionMenuProps}
                />
            </div>
        </div>
    )
}

LayerToolbar.propTypes = {
    toggleLayerVisibility: PropTypes.func.isRequired,
    onOpacityChange: PropTypes.func.isRequired,
    hasError: PropTypes.bool,
    hasOpacity: PropTypes.bool,
    isVisible: PropTypes.bool,
    opacity: PropTypes.number,
    onEdit: PropTypes.func,
}

export default LayerToolbar
