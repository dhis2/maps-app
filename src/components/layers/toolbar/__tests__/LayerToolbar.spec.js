import React from 'react';
import { shallow } from 'enzyme';
import { LayerToolbar } from '../LayerToolbar';

describe('LayerToolbar', () => {
    const shallowRenderLayerToolbar = props => (
        shallow(
            <LayerToolbar
                classes={{}}
                opacity={0}
                isVisible={true}
                toggleLayerVisibility={() => null}
                onOpacityChange={() => null}
                {...props}
            />
        )
    )
    it('Should render only a visibility toggle and opacity slider', () => {
        const wrapper = shallowRenderLayerToolbar();
        expect(wrapper.find('WithStyles(IconButton)').length).toBe(1); // Visibility toggle
        expect(wrapper.find('WithStyles(OpacitySlider)').length).toBe(1);
        expect(wrapper.find('WithStyles(LayerToolbarMoreMenu)').length).toBe(1);
    });

    it('Should show VisibilityIcon when visible', () => {
        const wrapper = shallowRenderLayerToolbar({
            isVisible: true,
        });
        expect(wrapper.find('pure(VisibilityIcon)').length).toBe(1);
        expect(wrapper.find('pure(VisibilityOffIcon)').length).toBe(0);
    });

    it('Should show VisibilityOffIcon when not visible', () => {
        const wrapper = shallowRenderLayerToolbar({
            isVisible: false,
        });
        expect(wrapper.find('pure(VisibilityIcon)').length).toBe(0);
        expect(wrapper.find('pure(VisibilityOffIcon)').length).toBe(1);
    });

    it('Should call toggleLayerVisibility callback on button press', () => {
        const toggleVisibleFn = jest.fn();
        const wrapper = shallowRenderLayerToolbar({
            toggleLayerVisibility: toggleVisibleFn,
        });
        wrapper.find('WithStyles(IconButton)').simulate('click');
        expect(toggleVisibleFn).toHaveBeenCalled();
    });

    it('Should match toolbar snapshot without Edit button', () => {
        const wrapper = shallowRenderLayerToolbar();
        expect(wrapper).toMatchSnapshot();
    });

    it('Should render edit button if passed onEdit prop', () => {
        const wrapper = shallowRenderLayerToolbar({
            onEdit: () => null,
        });
        expect(wrapper.find('WithStyles(IconButton)').length).toBe(2); // Visibility toggle and Edit
        expect(wrapper.find('WithStyles(OpacitySlider)').length).toBe(1);
        expect(wrapper.find('WithStyles(LayerToolbarMoreMenu)').length).toBe(1);
    });

    it('Should match toolbar snapshot WITH Edit button', () => {
        const wrapper = shallowRenderLayerToolbar({
            onEdit: () => null,
        })
        expect(wrapper).toMatchSnapshot();
    });

    it('Should call onEdit callback on button press', () => {
        const toggleVisibleFn = jest.fn();
        const editFn = jest.fn();

        const wrapper = shallowRenderLayerToolbar({
            toggleLayerVisibility: toggleVisibleFn,
            onEdit: editFn,
        });

        // First button should be Edit
        wrapper
            .find('WithStyles(IconButton)')
            .first()
            .simulate('click');
        expect(editFn).toHaveBeenCalled();
        expect(toggleVisibleFn).not.toHaveBeenCalled();

        // Second button should be Visibility Toggle
        wrapper
            .find('WithStyles(IconButton)')
            .at(1)
            .simulate('click');
        expect(toggleVisibleFn).toHaveBeenCalled();
        expect(editFn).toHaveBeenCalledTimes(1);
    });
});