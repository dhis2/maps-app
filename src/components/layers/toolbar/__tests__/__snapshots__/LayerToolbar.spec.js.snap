// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LayerToolbar Should match toolbar snapshot WITH Edit button 1`] = `
<WithStyles(Toolbar)>
  <WithStyles(Tooltip)
    key="edit"
    title="Edit"
  >
    <WithStyles(IconButton)
      onClick={[Function]}
    >
      <pure(CreateIcon) />
    </WithStyles(IconButton)>
  </WithStyles(Tooltip)>
  <span />
  <WithStyles(Tooltip)
    title="Toggle visibility"
  >
    <WithStyles(IconButton)
      onClick={[Function]}
    >
      <pure(VisibilityIcon) />
    </WithStyles(IconButton)>
  </WithStyles(Tooltip)>
  <div>
    <WithStyles(Tooltip)
      title="Set layer opacity"
    >
      <div>
        <WithStyles(OpacitySlider)
          classes={
            Object {
              "root": undefined,
            }
          }
          onChange={[Function]}
          opacity={0}
        />
      </div>
    </WithStyles(Tooltip)>
  </div>
  <WithStyles(LayerToolbarMoreMenu)
    classes={
      Object {
        "button": undefined,
      }
    }
    onEdit={[Function]}
  />
</WithStyles(Toolbar)>
`;

exports[`LayerToolbar Should match toolbar snapshot without Edit button 1`] = `
<WithStyles(Toolbar)>
  <WithStyles(Tooltip)
    title="Toggle visibility"
  >
    <WithStyles(IconButton)
      onClick={[Function]}
    >
      <pure(VisibilityIcon) />
    </WithStyles(IconButton)>
  </WithStyles(Tooltip)>
  <div>
    <WithStyles(Tooltip)
      title="Set layer opacity"
    >
      <div>
        <WithStyles(OpacitySlider)
          classes={
            Object {
              "root": undefined,
            }
          }
          onChange={[Function]}
          opacity={0}
        />
      </div>
    </WithStyles(Tooltip)>
  </div>
  <WithStyles(LayerToolbarMoreMenu)
    classes={
      Object {
        "button": undefined,
      }
    }
  />
</WithStyles(Toolbar)>
`;
